models/DataObject/AbstractObject.php

```diff
@@ -438,15 +441,14 @@ public function getChildren($objectTypes = [self::OBJECT_TYPE_OBJECT, self::OBJE
      */
     public function hasChildren($objectTypes = [self::OBJECT_TYPE_OBJECT, self::OBJECT_TYPE_FOLDER], $unpublished = false)
     {
-        if (is_bool($this->o_hasChildren)) {
-            if (($this->o_hasChildren and empty($this->o_children)) or (!$this->o_hasChildren and !empty($this->o_children))) {
-                return $this->getDao()->hasChildren($objectTypes, $unpublished);
-            } else {
-                return $this->o_hasChildren;
+        if (isset($this->hasChildren[$unpublished])) {
```

Comment:
We need a cache key considering the `$objectTypes` here as well. 
Use the to be created function `getListingCacheKey()` here.